addons/web/static/src/views/fields/translation_dialog.js

```diff
@@ -15,53 +16,53 @@ export class TranslationDialog extends Component {
-            const [translations, context] = await this.loadTranslations(languages);
+            const [translations, context] = await this.loadTranslations();
             let id = 1;
             translations.forEach((t) => (t.id = id++));
-            this.props.isText = context.translation_type === "text";
-            this.props.showSource = context.translation_show_source;
+            this.props.languages = await loadLanguages(this.orm);
+            if (!this.props.languages.find((l) => l[0] === "en_US")) {
+                this.props.languages.push(["en_US", _t("Source Value")]);
+            }
+            this.props.isText = ["text", "html"].includes(context.field_type);
+            this.props.translateType = context.translate_type;
 
-            this.terms = translations.map((term) => {
-                const relatedLanguage = languages.find((l) => l[0] === term.lang);
-                const termInfo = {
-                    ...term,
-                    langName: relatedLanguage[1],
-                    value: term.value || "",
-                };
-                // we set the translation value coming from the database, except for the language
-                // the user is currently utilizing. Then we set the translation value coming
-                // from the value of the field in the form
-                if (
-                    term.lang === this.user.lang &&
-                    !this.props.showSource &&
-                    !this.props.isComingFromTranslationAlert
-                ) {
-                    this.updatedTerms[term.id] = this.props.userLanguageValue;
-                    termInfo.value = this.props.userLanguageValue;
-                }
-                return termInfo;
-            });
-            this.terms.sort((a, b) => a.langName.localeCompare(b.langName));
+            this.state.terms.push(
+                ...translations.map((term) => {
+                    const relatedLanguage = this.props.languages.find((l) => l[0] === term.lang);
+                    const termInfo = {
+                        ...term,
+                        isTranslated: term.is_translated,
+                        langName: relatedLanguage[1],
+                        oldValue: term.value,
+                        isModified: false,
+                    };
+                    delete termInfo.is_translated;
+                    // we set the translation value coming from the database, except for the language
+                    // the user is currently utilizing. Then we set the translation value coming
+                    // from the value of the field in the form
+                    if (
+                        term.lang === this.user.lang &&
+                        this.props.translateType === "model" &&
+                        !this.props.isComingFromTranslationAlert
+                    ) {
+                        termInfo.value = this.props.userLanguageValue;
```

Comment:
While debugging, I noticed that `this.props.userLanguageValue` can be a `markup` object. Have you taken that into consideration?

Reply:
This logic is from long time ago. Could you tell me when the `this.props.userLanguageValue` can be a `markup` object(which field)? I will test if it has a bug